lib/dal/src/attribute/value/default_subscription.rs

```diff
@@ -0,0 +1,244 @@
+    let mut work_queue = VecDeque::from(root_children);
+    while let Some(dest_av_id) = work_queue.pop_front() {
+        let dest_prop_id = AttributeValue::prop_id(ctx, dest_av_id).await?;
+        let prop = Prop::get_by_id(ctx, dest_prop_id).await?;
+        let mut potential = None;
+
+        let children = AttributeValue::child_av_ids(ctx, dest_av_id).await?;
+        work_queue.extend(children);
+
+        if let Some(suggestions) = suggestions.get(&dest_prop_id) {
+            for &suggestion in suggestions {
+                if suggestion.dest_prop_id != dest_prop_id {
+                    continue;
+                }
+
+                let Some((source_prop, source_av_id)) =
+                    potential_source_props.get(&suggestion.source_prop_id)
+                else {
+                    continue;
+                };
+
+                if source_prop.kind == prop.kind {
+                    if potential.is_some() {
+                        // Ambiguous match!
+                        potential = None;
+                        break;
+                    }
+
+                    potential = Some(DefaultSubscription {
+                        source_av_id: *source_av_id,
+                        dest_av_id,
+                    });
+                }
+            }
+        }
+
+        if let Some(potential) = potential {
+            result.push(potential);
+        }
+    }
+
+    Ok(result)
+}
+
+/// Returns a map of potential *destination* prop ids to the
+/// ResolvedPropSuggestion for the entire change set
+pub async fn calculate_all_prop_suggestions_for_change_set(
+    ctx: &DalContext,
+) -> SchemaVariantResult<BTreeMap<PropId, BTreeSet<ResolvedPropSuggestion>>> {
+    let mut result: BTreeMap<PropId, BTreeSet<ResolvedPropSuggestion>> = BTreeMap::new();
+
+    for schema_id in Schema::list_ids(ctx).await? {
+        for variant_id in Schema::list_schema_variant_ids(ctx, schema_id).await? {
+            let mut suggestions = vec![];
+            suggestions.extend(SchemaVariant::props_suggested_as_sources(ctx, variant_id).await?);
+            suggestions
+                .extend(SchemaVariant::props_suggested_as_destinations(ctx, variant_id).await?);
+
+            for suggestion in suggestions {
+                result
+                    .entry(suggestion.dest_prop_id)
+                    .and_modify(|sugs| {
+                        sugs.insert(suggestion);
+                    })
+                    .or_insert_with(|| BTreeSet::from([suggestion]));
+            }
+        }
+    }
+
+    Ok(result)
+}
+
+impl AttributeValue {
+    pub async fn set_as_default_subscription_source(
+        ctx: &DalContext,
+        id: AttributeValueId,
+    ) -> AttributeValueResult<()> {
+        let snapshot = ctx.workspace_snapshot()?;
+
+        let category_id = get_or_create_default_subscription_category(ctx).await?;
+
+        snapshot
+            .add_edge(
```

Comment:
have we decided to stop using the `implement_add_edge_to!` macros? It felt like they made understanding nodes easier at a glance

Reply:
Good call